[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.7331351.svg)](https://doi.org/10.5281/zenodo.7331351)
[![DOI](https://joss.theoj.org/papers/10.21105/joss.04677/status.svg)](https://doi.org/10.21105/joss.04677)
[![Nextflow](https://img.shields.io/badge/nextflow%20DSL2-%E2%89%A521.04.0-23aa62.svg?labelColor=000000)](https://www.nextflow.io/)
[![run with conda](http://img.shields.io/badge/run%20with-conda-3EB049?labelColor=000000&logo=anaconda)](https://docs.conda.io/en/latest/)
[![run with docker](https://img.shields.io/badge/run%20with-docker-0db7ed?labelColor=000000&logo=docker)](https://www.docker.com/)
[![run with singularity](https://img.shields.io/badge/run%20with-singularity-1d355c.svg?labelColor=000000)](https://sylabs.io/docs/)
![License](https://img.shields.io/github/license/NorwegianVeterinaryInstitute/ALPPACA)

<p align="center"><img src="logo.png" alt="Ellipsis" width="600"></p>
Logo generated with www.biorender.com

# Information
ALPPACA is a pipeline for various phylogenetic analyses of prokaryotic isolates. 
The pipeline includes core genome-, core gene-, and reference-based (mapping) phylogeny, as well as FastANI and cgMLST (under development). 
Please see the [description](https://github.com/NorwegianVeterinaryInstitute/ALPPACA/wiki/1.-Pipeline-and-program-descriptions) page for more information.

# Installation and Running
Please check the [wiki](https://github.com/NorwegianVeterinaryInstitute/ALPPACA/wiki).
- [Installation](https://github.com/NorwegianVeterinaryInstitute/ALPPACA/wiki/2.-Installation)
- [How to run](https://github.com/NorwegianVeterinaryInstitute/ALPPACA/wiki/3.-Input-and-usage)
- [Output file descriptions](https://github.com/NorwegianVeterinaryInstitute/ALPPACA/wiki/5.-Output-files)

# Citation
ALPPACA has been published in Journal of Open Source Software. Use the following citation:

Kaspersen et al., (2022). ALPPACA - A tooL for Prokaryotic Phylogeny And Clustering Analysis. Journal of Open Source Software, 7(79), 4677, https://doi.org/10.21105/joss.04677
