# PhyloFlow
Nextflow pipeline for phylogenetic analysis.

## Disclaimer
This pipeline is currently under development. Contact Håkon Kaspersen for more information.

## Tracks

![pipeline](pipeline.png)

The pipeline consist of three tracks:

### CORE
This track identifies the core genes in the input assemblies by running Prokka [1],
followed by Panaroo QC and Panaroo pan-genome [2]. Then, duplicated genomes are removed with seqkit [3] (optional).
Then, a maximum likelihood (ML) tree is generated with IQTree [4], and SNP distances calculated with snp-dists [5].

### ASSEMBLY
In this track, the core genome is identified using ParSNP [6]. Then, duplicated genomes are removed with seqkit (optional).
Then, recombinant areas are identified with Gubbins [7], and masked with maskrc-svg [8] before the ML tree is generated 
with IQTree, and SNP-distances calculated with snp-dists. 

### READS
In this track, the core genome is generated through Snippy [9], which use reads
as input. Here too, recombinant areas are identified with Gubbins and subsequently masked with maskrc-svg,
and the tree generated by IQTree. SNP-distances are calculated with snp-dists.

## Usage
Copy the main.config file to your workspace and make necessary changes.
Then run the following:

```
module load Java/1.8.0_212

./PhyloFlow.sh /path/to/main.config /path/to/outdir
```

## Output files
Three output folders are generated, config_files, logs, and results.
Config_files contain the nextflow script and config files used in that specific run. The logs folder contain the logs of each program used in the pipeline. Finally, the files listed below is placed in the results folder.

### CORE

- **PANAROO_mashdist.txt**

   Mash distances, from PANAROO QC.

- **MDS_mash_plot.png**

   MDS plot of the mash results, from PANAROO QC. 

- **ncontigs_barplot.png**

   Number of contigs for each included genome, from PANAROO QC.

- **ngenes_barplot.png**

   Number of genes per included genome, from PANAROO PANGENOME.

- **PANAROO_pangenome_results.txt**

   Number of core genes, accessory genes and total genes, from PANAROO PANGENOME.

- **PANAROO_core_gene_alignment.aln**

   The concatenated core gene alignment of all included genomes in FASTA format, from PANAROO PANGENOME.

### ASSEMBLY

- **PARSNP_alignment.aln**

   The converted FASTA alignment of all included genomes, from PARSNP.

- **PARSNP_gingr_archive.ggr**

   Gingr-file used for visualizing the alignment, from PARSNP.

- **PARSNP_results.txt**

   Result file that contains information about the size (bp) and the percent coverage of each genome, from PARSNP. 

### READS

- **SNIPPY_alignment.aln**

   Alignment of all included genomes in FASTA format, from SNIPPY.

- **SNIPPY_results.txt**

   File that contains the result statistics, from SNIPPY.

### SHARED OUTPUT FILES
- **SEQKIT_deduplicated_alignment.fasta**

   Deduplicated alignment in FASTA format, from SEQKIT.

- **SEQKIT_deduplicated_sequences.fasta**

   The FASTA sequence of each duplicated genome, from SEQKIT.

- **SEQKIT_duplicated_list.txt**

   List of duplicated sequence IDs, from SEQKIT.

- **GUBBINS_filtered_alignment.aln** (not in CORE track)

   Alignment with recombinant sites removed in FASTA format, from GUBBINS. 

- **GUBBINS_statistics.txt** (not in CORE track)

   Text file with the result statistics, from GUBBINS.

- **MASKRC_masked_alignment.aln** (not in CORE track)

   Alignment with masked recombinant sites in FASTA format, from MASKRC.

- **MASKRC_recombinant_regions.txt** (not in CORE track)

   Text file with information about the recombinant regions, from MASKRC.

- **MASKRC_recombinant_plot.svg** (not in CORE track)

   Figure that shows the position of the recombinant areas, from MASKRC.

- **IQTREE_tree.phylo**

   Concensus tree with bootstrap values in NEXUS format, from IQTREE.

- **IQTREE_results.txt**

   Text file with results, from IQTREE.

- **SNPDIST_results.txt**

   The SNP distance matrix calculated from the alignment, from SNPDIST.


## References
[1] 10.1093/bioinformatics/btu153

[2] 10.1186/s13059-020-02090-4

[3] https://bioinf.shenwei.me/seqkit/

[4] 10.1093/molbev/msu300

[5] https://github.com/tseemann/snp-dists

[6] 10.1186/s13059-014-0524-x

[7] 10.1093/nar/gku1196

[8] https://github.com/kwongj/maskrc-svg

[9] https://github.com/tseemann/snippy






Håkon Kaspersen,
08.10.2020
