[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.7271324.svg)](https://doi.org/10.5281/zenodo.7271324)
![Platform](https://img.shields.io/badge/Platform-Nextflow-green)
![License](https://img.shields.io/github/license/NorwegianVeterinaryInstitute/ALPPACA)
[![status](https://joss.theoj.org/papers/182a5557ba7a30fe5c3ea732d2f1b222/status.svg)](https://joss.theoj.org/papers/182a5557ba7a30fe5c3ea732d2f1b222)

<p align="center"><img src="logo.png" alt="Ellipsis" width="600"></p>
Logo generated with www.biorender.com

# Information
ALPPACA is a pipeline for various phylogenetic analyses of prokaryotic isolates. 
The pipeline includes core genome-, core gene-, and reference-based (mapping) phylogeny. 
Please see the [description](https://github.com/NorwegianVeterinaryInstitute/ALPPACA/wiki/2.-Pipeline-and-program-descriptions) page for more information.


# Installation and Running
Please check the [wiki](https://github.com/NorwegianVeterinaryInstitute/ALPPACA/wiki).
- [Installation](https://github.com/NorwegianVeterinaryInstitute/ALPPACA/wiki/1.-Installation)
- [How to run](https://github.com/NorwegianVeterinaryInstitute/ALPPACA/wiki/3.-Input-and-usage)
- [Output file descriptions](https://github.com/NorwegianVeterinaryInstitute/ALPPACA/wiki/4.-Output-files)
